File: Sources/AIKit/OllamaProvider.swift
import Foundation

// MARK: - OllamaProvider

/// `AIProvider` for Ollama's local inference server.
///
/// Supports two API modes picked at runtime:
///   - `.native`       POST /api/chat  (Ollama-native JSON, stream=false)
///   - `.openaiCompat` POST /v1/chat/completions  (OpenAI-compat)
///
/// Endpoint validation: HTTP is allowed only for localhost / RFC-1918 ranges.
/// HTTPS is allowed for any host.
///
/// Multipart messages in native mode: text parts are joined; image parts are
/// sent as a top-level `images` array of base64 strings on the message object.
public final class OllamaProvider: AIProvider, Sendable {

    // MARK: - API mode

    public enum APIMode: String, Sendable {
        case native        // POST /api/chat
        case openaiCompat  // POST /v1/chat/completions
    }

    public typealias HTTPClient = @Sendable (URLRequest) async throws -> (Data, URLResponse)

    // MARK: - Stored

    private let endpoint: URL
    private let apiMode: APIMode
    private let bearerToken: String?
    private let httpClient: HTTPClient

    // MARK: - Init

    public init(
        endpoint: URL,
        apiMode: APIMode = .native,
        bearerToken: String? = nil,
        httpClient: @escaping HTTPClient = { try await URLSession.shared.data(for: $0) }
    ) throws {
        try Self.validate(endpoint: endpoint)
        self.endpoint = endpoint
        self.apiMode = apiMode
        self.bearerToken = bearerToken
        self.httpClient = httpClient
    }

    // MARK: - AIProvider

    public func chat(messages: [ChatMessage], config: AIConfig) async throws -> String {
        let request: URLRequest
        switch apiMode {
        case .native:
            request = try Self.buildNativeRequest(
                endpoint: endpoint,
                bearerToken: bearerToken,
                messages: messages,
                config: config
            )
        case .openaiCompat:
            request = try Self.buildOpenAICompatRequest(
                endpoint: endpoint,
                bearerToken: bearerToken,
                messages: messages,
                config: config
            )
        }

        let (data, response): (Data, URLResponse)
        do {
            (data, response) = try await httpClient(request)
        } catch {
            throw AIError.sendFailed(message: error.localizedDescription)
        }

        guard let httpResponse = response as? HTTPURLResponse else {
            throw AIError.sendFailed(message: "Non-HTTP response from Ollama")
        }

        switch httpResponse.statusCode {
        case 200:
            switch apiMode {
            case .native:       return try Self.parseNative(data: data)
            case .openaiCompat: return try Self.parseOpenAICompat(data: data)
            }
        case 401:
            throw AIError.authenticationFailed
        case 429:
            throw AIError.rateLimited
        default:
            let body = String(data: data, encoding: .utf8) ?? "<unreadable body>"
            throw AIError.sendFailed(message: "HTTP \(httpResponse.statusCode): \(body)")
        }
    }

    /// Lists available models via GET /api/tags. Used by Settings to populate the model picker.
    public func listModels() async throws -> [String] {
        let url = endpoint.appendingPathComponent("api/tags")
        var request = URLRequest(url: url)
        request.httpMethod = "GET"
        if let token = bearerToken {
            request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }

        let (data, response): (Data, URLResponse)
        do {
            (data, response) = try await httpClient(request)
        } catch {
            throw AIError.sendFailed(message: error.localizedDescription)
        }

        guard let httpResponse = response as? HTTPURLResponse, httpResponse.statusCode == 200 else {
            let body = String(data: data, encoding: .utf8) ?? ""
            throw AIError.sendFailed(message: "listModels failed: \(body)")
        }

        struct TagsResponse: Decodable {
            struct Model: Decodable { let name: String }
            let models: [Model]
        }
        do {
            let parsed = try JSONDecoder().decode(TagsResponse.self, from: data)
            return parsed.models.map { $0.name }
        } catch {
            throw AIError.decodingFailed(message: error.localizedDescription)
        }
    }

    // MARK: - Endpoint validation

    static func validate(endpoint: URL) throws {
        guard let scheme = endpoint.scheme?.lowercased() else { throw AIError.invalidEndpoint }
        if scheme == "https" { return }
        if scheme == "http" {
            guard let host = endpoint.host else { throw AIError.invalidEndpoint }
            if host == "localhost" || host == "127.0.0.1"
                || host.hasPrefix("10.")
                || host.hasPrefix("192.168.")
                || isPrivate172(host) {
                return
            }
            throw AIError.invalidEndpoint
        }
        throw AIError.invalidEndpoint
    }

    // MARK: - Native mode (/api/chat)

    static func buildNativeRequest(
        endpoint: URL,
        bearerToken: String?,
        messages: [ChatMessage],
        config: AIConfig
    ) throws -> URLRequest {
        let url = endpoint.appendingPathComponent("api/chat")
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "content-type")
        if let token = bearerToken {
            request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }

        // Prepend config.systemPrompt as a system message if present.
        var allMessages = messages
        if let systemPrompt = config.systemPrompt {
            allMessages.insert(ChatMessage(role: .system, content: systemPrompt), at: 0)
        }

        let serialized: [[String: Any]] = allMessages.map { msg -> [String: Any] in
            serializeNativeMessage(msg)
        }

        let body: [String: Any] = [
            "model": config.model,
            "messages": serialized,
            "stream": false,
            "options": [
                "temperature": config.temperature,
                "num_predict": config.maxTokens,
            ] as [String: Any],
        ]

        do {
            request.httpBody = try JSONSerialization.data(withJSONObject: body)
        } catch {
            throw AIError.sendFailed(message: "Could not serialize request body: \(error.localizedDescription)")
        }
        return request
    }

    /// Serialize a single message for the Ollama native /api/chat format.
    /// Text parts are joined; image parts become base64 strings in the `images` array.
    private static func serializeNativeMessage(_ msg: ChatMessage) -> [String: Any] {
        var textParts: [String] = []
        var imageBase64: [String] = []

        for part in msg.parts {
            switch part {
            case .text(let s):
                textParts.append(s)
            case .image(let jpegData, _):
                imageBase64.append(jpegData.base64EncodedString())
            }
        }

        var result: [String: Any] = [
            "role": msg.role.rawValue,
            "content": textParts.joined(separator: "\n"),
        ]
        if !imageBase64.isEmpty {
            result["images"] = imageBase64
        }
        return result
    }

    static func parseNative(data: Data) throws -> String {
        struct Response: Decodable {
            struct Message: Decodable { let content: String }
            let message: Message
        }
        do {
            let response = try JSONDecoder().decode(Response.self, from: data)
            return response.message.content
        } catch {
            throw AIError.decodingFailed(message: error.localizedDescription)
        }
    }

    // MARK: - OpenAI-compat mode (/v1/chat/completions)

    static func buildOpenAICompatRequest(
        endpoint: URL,
        bearerToken: String?,
        messages: [ChatMessage],
        config: AIConfig
    ) throws -> URLRequest {
        let url = endpoint.appendingPathComponent("v1/chat/completions")
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "content-type")
        // Use bearer token if set; otherwise omit Authorization header entirely.
        if let token = bearerToken {
            request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }

        // Prepend config.systemPrompt as a system message if present.
        var allMessages = messages
        if let systemPrompt = config.systemPrompt {
            allMessages.insert(ChatMessage(role: .system, content: systemPrompt), at: 0)
        }

        let body: [String: Any] = [
            "model": config.model,
            "max_tokens": config.maxTokens,
            "temperature": config.temperature,
            "messages": allMessages.map { msg -> [String: Any] in
                ["role": msg.role.rawValue, "content": serializeOpenAICompatParts(msg.parts)]
            },
        ]

        do {
            request.httpBody = try JSONSerialization.data(withJSONObject: body)
        } catch {
            throw AIError.sendFailed(message: "Could not serialize request body: \(error.localizedDescription)")
        }
        return request
    }

    /// Mirrors OpenAIProvider's serialization: single text → plain string; mixed → array of blocks.
    private static func serializeOpenAICompatParts(_ parts: [ChatMessage.Part]) -> Any {
        if parts.count == 1, case .text(let s) = parts[0] {
            return s
        }
        return parts.map { part -> [String: Any] in
            switch part {
            case .text(let s):
                return ["type": "text", "text": s]
            case .image(let jpegData, let mimeType):
                let dataURL = "data:\(mimeType);base64,\(jpegData.base64EncodedString())"
                return [
                    "type": "image_url",
                    "image_url": ["url": dataURL] as [String: Any],
                ]
            }
        }
    }

    static func parseOpenAICompat(data: Data) throws -> String {
        struct Response: Decodable {
            struct Choice: Decodable {
                struct Message: Decodable {
                    let role: String
                    let content: String
                }
                let message: Message
            }
            let choices: [Choice]
        }
        let response: Response
        do {
            response = try JSONDecoder().decode(Response.self, from: data)
        } catch {
            throw AIError.decodingFailed(message: error.localizedDescription)
        }
        guard let first = response.choices.first else {
            throw AIError.decodingFailed(message: "No choices in response")
        }
        return first.message.content
    }

    // MARK: - Private helpers

    /// Returns true for 172.16.x.x – 172.31.x.x (RFC-1918 range).
    private static func isPrivate172(_ host: String) -> Bool {
        let parts = host.split(separator: ".")
        guard parts.count == 4,
              parts[0] == "172",
              let second = Int(parts[1]),
              second >= 16, second <= 31 else { return false }
        return true
    }

    // Internal exposure for tests.
    static func isPrivate172Public(_ host: String) -> Bool { isPrivate172(host) }
}

File: windows/src/KosmoNotes.Providers/OllamaProvider.cs
using System.Net.Http;
using System.Net.Http.Headers;
using System.Text;
using System.Text.Json;
using System.Text.Json.Nodes;
using KosmoNotes.Core.Models;
using KosmoNotes.Providers.Internal;

namespace KosmoNotes.Providers;

/// <summary>
/// API mode selector for <see cref="OllamaProvider"/>. Mirrors the Swift
/// <c>OllamaProvider.APIMode</c> enum.
/// </summary>
public enum OllamaApiMode
{
    /// <summary>Native Ollama JSON over <c>POST /api/chat</c>.</summary>
    Native,

    /// <summary>OpenAI-compatible payload over <c>POST /v1/chat/completions</c>.</summary>
    OpenAICompat,
}

/// <summary>
/// <see cref="IAIProvider"/> for Ollama's local inference server.
/// </summary>
/// <remarks>
/// <para>Two API modes selected at construction time:</para>
/// <list type="bullet">
///   <item><see cref="OllamaApiMode.Native"/>: <c>POST /api/chat</c> — native Ollama
///         JSON, <c>stream:false</c>.</item>
///   <item><see cref="OllamaApiMode.OpenAICompat"/>: <c>POST /v1/chat/completions</c>
///         — OpenAI-compatible payload.</item>
/// </list>
/// <para>Endpoint validation: <c>https</c> is allowed for any host;
/// <c>http</c> only for <c>localhost</c>, <c>127.0.0.1</c>, and the RFC-1918
/// private ranges (<c>10.*</c>, <c>192.168.*</c>, <c>172.16.*–172.31.*</c>).
/// Throws <see cref="InvalidEndpointException"/> from the constructor on
/// any other endpoint.</para>
/// </remarks>
public sealed class OllamaProvider : IAIProvider
{
    private readonly Uri _endpoint;
    private readonly OllamaApiMode _apiMode;
    private readonly string? _bearerToken;
    private readonly HttpClient _httpClient;

    /// <summary>Create a provider against the given Ollama endpoint.</summary>
    /// <param name="endpoint">Base endpoint (e.g. <c>http://localhost:11434</c>).</param>
    /// <param name="apiMode">Which API surface to call.</param>
    /// <param name="bearerToken">Optional Authorization header value (rare for local Ollama).</param>
    /// <param name="httpClient">Optional client for tests.</param>
    /// <exception cref="InvalidEndpointException">If the endpoint is not allowed.</exception>
    public OllamaProvider(
        Uri endpoint,
        OllamaApiMode apiMode = OllamaApiMode.Native,
        string? bearerToken = null,
        HttpClient? httpClient = null)
    {
        ValidateEndpoint(endpoint);
        _endpoint = endpoint;
        _apiMode = apiMode;
        _bearerToken = bearerToken;
        _httpClient = httpClient ?? HttpDefaults.Default;
    }

    /// <inheritdoc />
    public async Task<string> ChatAsync(
        IReadOnlyList<ChatMessage> messages,
        AIConfig config,
        CancellationToken ct = default)
    {
        using var request = _apiMode switch
        {
            OllamaApiMode.Native =>
                BuildNativeRequest(_endpoint, _bearerToken, messages, config),
            OllamaApiMode.OpenAICompat =>
                BuildOpenAICompatRequest(_endpoint, _bearerToken, messages, config),
            _ => throw new ArgumentOutOfRangeException(nameof(_apiMode)),
        };

        HttpResponseMessage response;
        try
        {
            response = await _httpClient.SendAsync(request, ct).ConfigureAwait(false);
        }
        catch (HttpRequestException ex)
        {
            throw new SendFailedException(ex.Message, ex);
        }

        try
        {
            await HttpStatus.ThrowIfNotSuccessAsync(response, ct).ConfigureAwait(false);
            byte[] body = await response.Content.ReadAsByteArrayAsync(ct).ConfigureAwait(false);
            return _apiMode switch
            {
                OllamaApiMode.Native => ParseNative(body),
                OllamaApiMode.OpenAICompat => OpenAIProvider.Parse(body),
                _ => throw new ArgumentOutOfRangeException(),
            };
        }
        finally
        {
            response.Dispose();
        }
    }

    /// <summary>Lists available models via <c>GET /api/tags</c>.</summary>
    /// <param name="ct">Cancellation token.</param>
    public async Task<IReadOnlyList<string>> ListModelsAsync(CancellationToken ct = default)
    {
        var url = AppendPath(_endpoint, "api/tags");
        using var request = new HttpRequestMessage(HttpMethod.Get, url);
        if (_bearerToken is not null)
        {
            request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", _bearerToken);
        }

        HttpResponseMessage response;
        try
        {
            response = await _httpClient.SendAsync(request, ct).ConfigureAwait(false);
        }
        catch (HttpRequestException ex)
        {
            throw new SendFailedException(ex.Message, ex);
        }

        try
        {
            if (!response.IsSuccessStatusCode)
            {
                string body = await HttpStatus.ReadBodySafeAsync(response, ct).ConfigureAwait(false);
                throw new SendFailedException($"listModels failed: {body}");
            }

            byte[] data = await response.Content.ReadAsByteArrayAsync(ct).ConfigureAwait(false);
            try
            {
                using var doc = JsonDocument.Parse(data);
                if (!doc.RootElement.TryGetProperty("models", out var models) ||
                    models.ValueKind != JsonValueKind.Array)
                {
                    return Array.Empty<string>();
                }
                var result = new List<string>(models.GetArrayLength());
                foreach (var m in models.EnumerateArray())
                {
                    if (m.TryGetProperty("name", out var name) &&
                        name.ValueKind == JsonValueKind.String)
                    {
                        result.Add(name.GetString()!);
                    }
                }
                return result;
            }
            catch (JsonException ex)
            {
                throw new DecodingFailedException(ex.Message, ex);
            }
        }
        finally
        {
            response.Dispose();
        }
    }

    // ---- Endpoint validation ----------------------------------------------

    /// <summary>
    /// Validate that <paramref name="endpoint"/> is allowed (HTTPS anywhere;
    /// HTTP only for localhost / RFC-1918). Throws on failure.
    /// </summary>
    public static void ValidateEndpoint(Uri endpoint)
    {
        if (endpoint is null) throw new InvalidEndpointException();
        string scheme = endpoint.Scheme?.ToLowerInvariant() ?? string.Empty;
        if (scheme == "https") return;

        if (scheme != "http") throw new InvalidEndpointException();

        string host = endpoint.Host?.ToLowerInvariant() ?? string.Empty;
        if (string.IsNullOrEmpty(host)) throw new InvalidEndpointException();

        if (host == "localhost" || host == "127.0.0.1") return;
        if (host.StartsWith("10.", StringComparison.Ordinal)) return;
        if (host.StartsWith("192.168.", StringComparison.Ordinal)) return;
        if (IsPrivate172(host)) return;

        throw new InvalidEndpointException();
    }

    /// <summary>Returns true for hosts in the 172.16.0.0–172.31.255.255 range.</summary>
    internal static bool IsPrivate172(string host)
    {
        var parts = host.Split('.');
        if (parts.Length != 4) return false;
        if (parts[0] != "172") return false;
        if (!int.TryParse(parts[1], out int second)) return false;
        return second >= 16 && second <= 31;
    }

    // ---- Native request ---------------------------------------------------

    internal static HttpRequestMessage BuildNativeRequest(
        Uri endpoint,
        string? bearerToken,
        IReadOnlyList<ChatMessage> messages,
        AIConfig config)
    {
        var url = AppendPath(endpoint, "api/chat");
        var request = new HttpRequestMessage(HttpMethod.Post, url);
        if (bearerToken is not null)
        {
            request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", bearerToken);
        }

        var allMessages = new List<ChatMessage>(messages.Count + 1);
        if (config.SystemPrompt is { } systemPrompt)
        {
            allMessages.Add(ChatMessage.FromText(ChatRole.System, systemPrompt));
        }
        allMessages.AddRange(messages);

        var serialized = new JsonArray();
        foreach (var msg in allMessages)
        {
            serialized.Add(SerializeNativeMessage(msg));
        }

        var body = new JsonObject
        {
            ["model"] = config.Model,
            ["messages"] = serialized,
            ["stream"] = false,
            ["options"] = new JsonObject
            {
                ["temperature"] = config.Temperature,
                ["num_predict"] = config.MaxTokens,
            },
        };

        request.Content = new StringContent(body.ToJsonString(), Encoding.UTF8);
        request.Content.Headers.ContentType = new MediaTypeHeaderValue("application/json");
        return request;
    }

    private static JsonObject SerializeNativeMessage(ChatMessage msg)
    {
        var textParts = new List<string>();
        var images = new List<string>();
        foreach (var part in msg.Parts)
        {
            switch (part)
            {
                case TextPart t:
                    textParts.Add(t.Text);
                    break;
                case ImagePart img:
                    images.Add(Convert.ToBase64String(img.JpegData));
                    break;
            }
        }

        var obj = new JsonObject
        {
            ["role"] = RoleToString(msg.Role),
            ["content"] = string.Join("\n", textParts),
        };
        if (images.Count > 0)
        {
            var imgs = new JsonArray();
            foreach (var b64 in images) imgs.Add(b64);
            obj["images"] = imgs;
        }
        return obj;
    }

    internal static string ParseNative(byte[] data)
    {
        try
        {
            using var doc = JsonDocument.Parse(data);
            if (!doc.RootElement.TryGetProperty("message", out var message) ||
                !message.TryGetProperty("content", out var content))
            {
                throw new DecodingFailedException("Missing message.content in response.");
            }
            return content.GetString() ?? string.Empty;
        }
        catch (DecodingFailedException)
        {
            throw;
        }
        catch (JsonException ex)
        {
            throw new DecodingFailedException(ex.Message, ex);
        }
    }

    // ---- OpenAI-compat request --------------------------------------------

    internal static HttpRequestMessage BuildOpenAICompatRequest(
        Uri endpoint,
        string? bearerToken,
        IReadOnlyList<ChatMessage> messages,
        AIConfig config)
    {
        var url = AppendPath(endpoint, "v1/chat/completions");
        var request = new HttpRequestMessage(HttpMethod.Post, url);
        if (bearerToken is not null)
        {
            request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", bearerToken);
        }

        string json = OpenAIProvider.BuildBodyJson(messages, config);
        request.Content = new StringContent(json, Encoding.UTF8);
        request.Content.Headers.ContentType = new MediaTypeHeaderValue("application/json");
        return request;
    }

    // ---- Helpers ----------------------------------------------------------

    /// <summary>
    /// Append <paramref name="path"/> to <paramref name="baseUri"/> mirroring
    /// Swift's <c>URL.appendingPathComponent</c> — preserves the existing
    /// path and adds a single slash separator.
    /// </summary>
    internal static Uri AppendPath(Uri baseUri, string path)
    {
        string left = baseUri.GetLeftPart(UriPartial.Authority);
        string existingPath = baseUri.AbsolutePath;
        if (!existingPath.EndsWith('/')) existingPath += "/";
        // Trim any leading slash on the suffix to avoid double-slash.
        string suffix = path.TrimStart('/');
        return new Uri(left + existingPath + suffix);
    }

    private static string RoleToString(ChatRole role) => role switch
    {
        ChatRole.System => "system",
        ChatRole.User => "user",
        ChatRole.Assistant => "assistant",
        _ => throw new ArgumentOutOfRangeException(nameof(role)),
    };
}
